pinot-core/src/main/java/org/apache/pinot/core/data/readers/JSONRecordReader.java

```diff
@@ -71,25 +73,19 @@ public GenericRow next() {
     return next(new GenericRow());
   }
 
+  // NOTE: hard to extract common code further
+  @SuppressWarnings("Duplicates")
   @Override
   public GenericRow next(GenericRow reuse) {
     Map record = _iterator.next();
-
-    for (FieldSpec fieldSpec : _schema.getAllFieldSpecs()) {
+    for (FieldSpec fieldSpec : _fieldSpecs) {
       String fieldName = fieldSpec.getName();
-      Object jsonValue = record.get(fieldName);
-
-      Object value;
-      if (fieldSpec.isSingleValueField()) {
-        String token = jsonValue != null ? jsonValue.toString() : null;
-        value = RecordReaderUtils.convertToDataType(token, fieldSpec);
-      } else {
-        value = RecordReaderUtils.convertToDataTypeArray((ArrayList) jsonValue, fieldSpec);
+      Object value = record.get(fieldName);
+      // Allow default value for non-time columns
+      if (value != null || fieldSpec.getFieldType() != FieldSpec.FieldType.TIME) {
```

Comment:
We cannot throw exception because it is valid if only one of incoming or outgoing column is not in the record. The time converter can catch this and throw exception if no time column is set.